File: dotnet/src/SemanticKernel.Core/CoreSkills/ConversationSummarySkill.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.SemanticKernel.Orchestration;
using Microsoft.SemanticKernel.SemanticFunctions.Partitioning;
using Microsoft.SemanticKernel.SkillDefinition;

namespace Microsoft.SemanticKernel.CoreSkills;

/// <summary>
/// <para>Semantic skill that enables conversations summarization.</para>
/// </summary>
/// <example>
/// <code>
/// var kernel Kernel.Builder.Build();
/// kernel.ImportSkill(new ConversationSummarySkill(kernel));
/// </code>
/// </example>
public class ConversationSummarySkill
{
    /// <summary>
    /// The max tokens to process in a single semantic function call.
    /// </summary>
    private const int MaxTokens = 1024;

    private readonly ISKFunction _summarizeConversationFunction;
    private readonly ISKFunction _conversationActionItemsFunction;
    private readonly ISKFunction _conversationTopicsFunction;

    /// <summary>
    /// Initializes a new instance of the <see cref="ConversationSummarySkill"/> class.
    /// </summary>
    /// <param name="kernel">Kernel instance</param>
    public ConversationSummarySkill(IKernel kernel)
    {
        this._summarizeConversationFunction = kernel.CreateSemanticFunction(
            SemanticFunctionConstants.SummarizeConversationDefinition,
            skillName: nameof(ConversationSummarySkill),
            description: "Given a section of a conversation transcript, summarize the part of the conversation.",
            maxTokens: MaxTokens,
            temperature: 0.1,
            topP: 0.5);

        this._conversationActionItemsFunction = kernel.CreateSemanticFunction(
            SemanticFunctionConstants.GetConversationActionItemsDefinition,
            skillName: nameof(ConversationSummarySkill),
            description: "Given a section of a conversation transcript, identify action items.",
            maxTokens: MaxTokens,
            temperature: 0.1,
            topP: 0.5);

        this._conversationTopicsFunction = kernel.CreateSemanticFunction(
            SemanticFunctionConstants.GetConversationTopicsDefinition,
            skillName: nameof(ConversationSummarySkill),
            description: "Analyze a conversation transcript and extract key topics worth remembering.",
            maxTokens: MaxTokens,
            temperature: 0.1,
            topP: 0.5);
    }

    /// <summary>
    /// Given a long conversation transcript, summarize the conversation.
    /// </summary>
    /// <param name="input">A long conversation transcript.</param>
    /// <param name="context">The SKContext for function execution.</param>
    [SKFunction("Given a long conversation transcript, summarize the conversation.")]
    [SKFunctionName("SummarizeConversation")]
    [SKFunctionInput(Description = "A long conversation transcript.")]
    public Task<SKContext> SummarizeConversationAsync(string input, SKContext context)
    {
        List<string> lines = SemanticTextPartitioner.SplitPlainTextLines(input, MaxTokens);
        List<string> paragraphs = SemanticTextPartitioner.SplitPlainTextParagraphs(lines, MaxTokens);

        return this._summarizeConversationFunction
            .AggregatePartitionedResultsAsync(paragraphs, context);
    }

    /// <summary>
    /// Given a long conversation transcript, identify action items.
    /// </summary>
    /// <param name="input">A long conversation transcript.</param>
    /// <param name="context">The SKContext for function execution.</param>
    [SKFunction("Given a long conversation transcript, identify action items.")]
    [SKFunctionName("GetConversationActionItems")]
    [SKFunctionInput(Description = "A long conversation transcript.")]
    public Task<SKContext> GetConversationActionItemsAsync(string input, SKContext context)
    {
        List<string> lines = SemanticTextPartitioner.SplitPlainTextLines(input, MaxTokens);
        List<string> paragraphs = SemanticTextPartitioner.SplitPlainTextParagraphs(lines, MaxTokens);

        return this._conversationActionItemsFunction
            .AggregatePartitionedResultsAsync(paragraphs, context);
    }

    /// <summary>
    /// Given a long conversation transcript, identify topics.
    /// </summary>
    /// <param name="input">A long conversation transcript.</param>
    /// <param name="context">The SKContext for function execution.</param>
    [SKFunction("Given a long conversation transcript, identify topics worth remembering.")]
    [SKFunctionName("GetConversationTopics")]
    [SKFunctionInput(Description = "A long conversation transcript.")]
    public Task<SKContext> GetConversationTopicsAsync(string input, SKContext context)
    {
        List<string> lines = SemanticTextPartitioner.SplitPlainTextLines(input, MaxTokens);
        List<string> paragraphs = SemanticTextPartitioner.SplitPlainTextParagraphs(lines, MaxTokens);

        return this._conversationTopicsFunction
            .AggregatePartitionedResultsAsync(paragraphs, context);
    }
}

File: java/plugins/semantickernel-plugin-core/src/main/java/com/microsoft/semantickernel/coreskills/ConversationSummarySkill.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.coreskills;

import com.microsoft.semantickernel.Kernel;
import com.microsoft.semantickernel.SKBuilders;
import com.microsoft.semantickernel.orchestration.SKContext;
import com.microsoft.semantickernel.skilldefinition.annotations.DefineSKFunction;
import com.microsoft.semantickernel.skilldefinition.annotations.SKFunctionInputAttribute;
import com.microsoft.semantickernel.text.TextChunker;
import com.microsoft.semantickernel.textcompletion.CompletionSKFunction;
import java.util.List;
import javax.annotation.Nullable;
import reactor.core.publisher.Mono;

/** Semantic skill that enables conversations summarization */
public class ConversationSummarySkill {
    // The max tokens to process in a single semantic function call.
    private static final int MaxTokens = 1024;

    private final CompletionSKFunction summarizeConversationFunction;
    private final CompletionSKFunction conversationActionItemsFunction;
    private final CompletionSKFunction conversationTopicsFunction;
    private final Kernel kernel;

    /**
     * Initializes a new instance of the ConversationSummarySkill class
     *
     * @param kernel Kernel instance
     */
    public ConversationSummarySkill(Kernel kernel) {
        this.kernel = kernel;
        this.summarizeConversationFunction =
                SKBuilders.completionFunctions()
                        .withKernel(kernel)
                        .withPromptTemplate(
                                SemanticFunctionConstants.SummarizeConversationDefinition)
                        .withFunctionName("summarize")
                        .withSkillName("ConversationSummarySkill")
                        .withDescription(
                                "Given a section of a conversation transcript, summarize the part"
                                        + " of the conversation.")
                        .withCompletionConfig(
                                SKBuilders.completionConfig()
                                        .maxTokens(MaxTokens)
                                        .temperature(0.1)
                                        .topP(0.5)
                                        .build())
                        .build();

        this.conversationActionItemsFunction =
                SKBuilders.completionFunctions()
                        .withKernel(kernel)
                        .withPromptTemplate(
                                SemanticFunctionConstants.GetConversationActionItemsDefinition)
                        .withFunctionName("ActionItems")
                        .withSkillName("ConversationSummarySkill")
                        .withDescription(
                                "Given a section of a conversation transcript, identify action"
                                        + " items.")
                        .withCompletionConfig(
                                SKBuilders.completionConfig()
                                        .maxTokens(MaxTokens)
                                        .temperature(0.1)
                                        .topP(0.5)
                                        .build())
                        .build();

        this.conversationTopicsFunction =
                SKBuilders.completionFunctions()
                        .withKernel(kernel)
                        .withPromptTemplate(
                                SemanticFunctionConstants.GetConversationTopicsDefinition)
                        .withFunctionName("Topics")
                        .withSkillName("ConversationSummarySkill")
                        .withDescription(
                                "Analyze a conversation transcript and extract key topics worth"
                                        + " remembering.")
                        .withCompletionConfig(
                                SKBuilders.completionConfig()
                                        .maxTokens(MaxTokens)
                                        .temperature(0.1)
                                        .topP(0.5)
                                        .build())
                        .build();
    }

    /**
     * Given a long conversation transcript, summarize the conversation
     *
     * @param input A long conversation transcript
     * @param context The SKContext for function execution
     * @return A context containing the summarized conversation
     */
    @DefineSKFunction(
            description = "Given a long conversation transcript, summarize the conversation.",
            name = "SummarizeConversation")
    public Mono<SKContext> summarizeConversationAsync(
            @SKFunctionInputAttribute(description = "A long conversation transcript.") String input,
            @Nullable SKContext context) {
        List<String> lines = TextChunker.splitPlainTextLines(input, MaxTokens);
        List<String> paragraphs = TextChunker.splitPlainTextParagraphs(lines, MaxTokens);

        if (context == null) {
            context = SKBuilders.context().withSkills(kernel.getSkills()).build();
        }

        SKContext completionContext = context.copy();

        return this.summarizeConversationFunction.aggregatePartitionedResultsAsync(
                paragraphs, completionContext);
    }

    /**
     * Given a long conversation transcript, identify action items
     *
     * @param input A long conversation transcript
     * @param context The SKContext for function execution
     * @return A context containing the action items
     */
    @DefineSKFunction(
            description = "Given a long conversation transcript, identify action items.",
            name = "GetConversationActionItems")
    public Mono<SKContext> getConversationActionItemsAsync(
            @SKFunctionInputAttribute(description = "A long conversation transcript.") String input,
            SKContext context) {
        List<String> lines = TextChunker.splitPlainTextLines(input, MaxTokens);
        List<String> paragraphs = TextChunker.splitPlainTextParagraphs(lines, MaxTokens);

        SKContext completionContext = context.copy();

        return this.conversationActionItemsFunction.aggregatePartitionedResultsAsync(
                paragraphs, completionContext);
    }

    /**
     * Given a long conversation transcript, identify topics
     *
     * @param input A long conversation transcript
     * @param context The SKContext for function execution
     * @return A context containing the topics
     */
    @DefineSKFunction(
            description =
                    "Given a long conversation transcript, identify topics worth remembering.",
            name = "GetConversationTopics")
    public Mono<SKContext> getConversationTopicsAsync(
            @SKFunctionInputAttribute(description = "A long conversation transcript.") String input,
            SKContext context) {
        List<String> lines = TextChunker.splitPlainTextLines(input, MaxTokens);
        List<String> paragraphs = TextChunker.splitPlainTextParagraphs(lines, MaxTokens);

        SKContext completionContext = context.copy();

        return this.conversationTopicsFunction.aggregatePartitionedResultsAsync(
                paragraphs, completionContext);
    }
}
